File: src/Domain/Entities/Accountant/Category.cs
namespace Domain.Entities.Accountant;

public class Category : Entity
{
    public int Id { get; set; }
    public int? ParentId { get; set; }
    public int UserId { get; set; }
    public string Name { get; set; }
    public CategoryType Type { get; set; }
    public bool GenerateUpcomingExpense { get; set; }
    public bool IsTax { get; set; }
}

public enum CategoryType
{
    AllTransactions,
    DepositOnly,
    WithdrawalOnly
}

File: src/Clients/accountant/src/models/entities/category.ts
import { Syncable } from "models/sync/syncable";

export class Category implements Syncable {
  synced = false;
  parent: Category;

  constructor(
    public id: number,
    public parentId: number,
    public name: string,
    public type: CategoryType,
    public generateUpcomingExpense: boolean,
    public isTax: boolean,
    public createdDate: Date,
    public modifiedDate: Date
  ) {}

  get fullName(): string {
    return this.parent ? `${this.parent.name}/${this.name}` : this.name;
  }
}

export enum CategoryType {
  AllTransactions,
  DepositOnly,
  ExpenseOnly,
}
